storage-proofs/porep/src/nse/circuit/hash.rs

```diff
@@ -0,0 +1,276 @@
+    mut cs: CS,
+    replica_id: &[Boolean],
+    challenge_index_num: &UInt64,
+    layer: u32,
+    config: &Config,
+    parents_data: &[num::AllocatedNum<Bls12>],
+) -> Result<num::AllocatedNum<Bls12>, SynthesisError> {
+    // TODO: can we reuse this allocation accross challenges?
+    let layer_index = UInt32::alloc(cs.namespace(|| "layer_index"), Some(layer))?;
+
+    // prefix
+    let mut ciphertexts = hash_prefix(challenge_index_num, layer_index)?;
+
+    // replica id
+    ciphertexts.extend_from_slice(replica_id);
+
+    // batch parents
+    ciphertexts.extend_from_slice(&batch_expansion(
+        cs.namespace(|| "batch_expansion"),
+        config.k as usize,
+        config.degree_expander,
+        parents_data,
+    )?);
+
+    // sha256
+    hash_sha256(cs, &ciphertexts)
+}
+
+pub fn derive_butterfly_layer_leaf<CS: ConstraintSystem<Bls12>>(
+    mut cs: CS,
+    replica_id: &[Boolean],
+    challenge_index_num: &UInt64,
+    layer: u32,
+    parents_data: &[num::AllocatedNum<Bls12>],
+) -> Result<num::AllocatedNum<Bls12>, SynthesisError> {
+    // TODO: can we reuse this allocation accross challenges?
+    let layer_index = UInt32::alloc(cs.namespace(|| "layer_index"), Some(layer))?;
+
+    // prefix
+    let mut ciphertexts = hash_prefix(challenge_index_num, layer_index)?;
+
+    // replica id
+    ciphertexts.extend_from_slice(replica_id);
+
+    // hash parents
+    ciphertexts.extend(expand_nums_to_bools(
+        cs.namespace(|| "expand_nums"),
+        parents_data,
+    )?);
+
+    // sha256
+    hash_sha256(cs, &ciphertexts)
+}
+
+pub fn derive_last_layer_leaf<CS: ConstraintSystem<Bls12>>(
+    mut cs: CS,
+    replica_id: &[Boolean],
+    challenge_index_num: &UInt64,
+    layer: u32,
+    data_node: &num::AllocatedNum<Bls12>,
+    parents_data: &[num::AllocatedNum<Bls12>],
+) -> Result<num::AllocatedNum<Bls12>, SynthesisError> {
+    // derive regular label
+    let key = derive_butterfly_layer_leaf(
+        cs.namespace(|| "butterfly_label"),
+        replica_id,
+        challenge_index_num,
+        layer,
+        parents_data,
+    )?;
+
+    // encode
+    encode::encode(cs.namespace(|| "encode"), &key, data_node)
+}
+
+/// Sha256
+fn hash_sha256<CS: ConstraintSystem<Bls12>>(
+    mut cs: CS,
+    ciphertexts: &[Boolean],
+) -> Result<num::AllocatedNum<Bls12>, SynthesisError> {
+    // Compute Sha256
+    let alloc_bits = sha256_circuit(cs.namespace(|| "sha256"), &ciphertexts[..])?;
+
+    // Convert the hash result into a single Fr.
+    let bits = reverse_bit_numbering(alloc_bits);
+    multipack::pack_bits(
+        cs.namespace(|| "sha256_result_num"),
+        &bits[0..(Fr::CAPACITY as usize)],
+    )
+}
+
+/// Calculates the common prefix for the hash.
+fn hash_prefix(challenge_index: &UInt64, layer: UInt32) -> Result<Vec<Boolean>, SynthesisError> {
+    // The prefix is 32 bytes long
+    const PREFIX_LEN: usize = 32 * 8;
+
+    let mut ciphertexts = Vec::with_capacity(PREFIX_LEN);
```

Comment:
Why is this called `ciphertexts`?

Reply:
legacy, and because I like it as a name for sth encoded